staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go

```diff
@@ -29,11 +29,22 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/json"
+
+	"github.com/golang/glog"
 )
 
-// NestedField returns the value of a nested field.
+// NestedField returns a deep copy of the value of a nested field.
 // false is returned if the value is missing.
+// nil, true is returned for a nil field.
 func NestedField(obj map[string]interface{}, fields ...string) (interface{}, bool) {
+	val, ok := nestedFieldNoCopy(obj, fields...)
+	if !ok {
+		return nil, false
+	}
+	return unstructured.DeepCopyJSONValue(val), true
```

Comment:
why are we deep copying here? need to double check this doesn't impact the uses by patch. cc @wojtek-t

Reply:
I think it probably will impact.

Before proceeding with this change (in general with this PR), I would like to see the impact of those changes to this benchmark:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unstructured_test.go#L134

Reply:
Deep copying only matters (and is done in this PR) for slices and maps. Both `getNestedSlice()` and `getNestedMap()` in `master` currently make a copy: 
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L209-L233
Another place where shallow copying is done currently is `getOwnerReferences()` - a helper function used by corresponding methods on Unstructured(List):
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L333-L356
This is actually unnecessary because `GetOwnerReferences()` method converts data into structs using `extractOwnerReference ()`:
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L358-L369
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L285-L313

I have removed this copying.
Regarding the benchmark - looks like there is no difference (probably because there is not much extra copying happening):
```
on master
/usr/local/Cellar/go/1.9.1/libexec/bin/go test -c -i -o /private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go k8s.io/kubernetes/pkg/api/testing
/private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go -test.v -test.bench ^BenchmarkToFromUnstructured$ -test.run ^$
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/api/testing
20000	     79458 ns/op
PASS

Process finished with exit code 0
```
```
this PR:
/usr/local/Cellar/go/1.9.1/libexec/bin/go test -c -i -o /private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go k8s.io/kubernetes/pkg/api/testing
/private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go -test.v -test.bench ^BenchmarkToFromUnstructured$ -test.run ^$
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/api/testing
20000	     79527 ns/op
PASS

Process finished with exit code 0
```
Actually, I have reviewed usages of this method and replaced two places where it was used with unexported `nestedFieldNoCopy()`. So this change should have 0 impact.